### PR TITLE
Imdbid from nfo and sync / sorting

### PR DESCRIPTION
**Pull request by goodjobswe**
Added to sort the most likely wanted subtitles on top even if moviehash search fails, added another condition to mark the subtitle as in sync.

Added support for fetching IMDB id from nfo.

Changed so that if a manual string is entered, that string is the only thing being searched for. Don't append the query string to the search if searching by moviehash or IMDB id succeeded, it returns to many unwanted results. Only use the query string in the search if moviehash and IMDB id search returns empty or if manual search or if it's a TV Show.

### CODE REVIEW

**Comment by goodjobswe** [comment]
Any comments on the latest changes?

File: service.subtitles.opensubtitles/service.py [hunk 1]
```diff
@@ -30,6 +31,11 @@
 
 from OSUtilities import OSDBServer, log, hashFile, normalizeString
 
+tree = ET.parse(xbmcvfs.File(os.path.dirname(os.path.realpath(__file__)).replace('addons','userdata').replace('service.subtitles.opensubtitles','guisettings.xml')))
+root = tree.getroot()
+for subtitlelanguage in root.iter('subtitlelanguage'):
+  PreferredSub = subtitlelanguage.text
+
```
Comment (amet) [inline service.subtitles.opensubtitles/service.py]: @fape did a nice PR https://github.com/xbmc/xbmc/pull/4458  where preferred language gets sent from xbmc. see here https://github.com/fape/service.subtitles.supersubtitles/commit/d64b3e13379cb754a6177b0b91665d08a8c40154

digging into guisettings is never a good idea

File: service.subtitles.opensubtitles/resources/lib/OSUtilities.py [hunk 2]
```diff
@@ -47,16 +47,32 @@ def searchsubtitles( self, item):
...
+      	search = self.server.SearchSubtitles( self.osdb_token, searchlist )
+      	if search["data"]:
+          return search["data"] 
+	else:
+	  searchlist.append({'sublanguageid':",".join(item['3let_language']),
+                	     'query'       :OS_search_string
+                            })
+      	  search = self.server.SearchSubtitles( self.osdb_token, searchlist )
+      	  if search["data"]:
+            return search["data"]
```
Comment (amet) [inline service.subtitles.opensubtitles/resources/lib/OSUtilities.py:75]: if its tested, fine

File: service.subtitles.opensubtitles/service.py (hunk 1 above)
Comment (goodjobswe) [inline service.subtitles.opensubtitles/service.py]: Yeah I know, just did it as a temporary solution before I knew if there where any other way of doing it, I will look into it.

**Comment by goodjobswe** [comment]
If I understand correct it is as simple as this? Since .get will return none if not found.

**Comment by amet** [comment]
if it works then yes, squash all commits to make it easy to review.

you can force push to your github branch and it will be updated here

**Comment by goodjobswe** [comment]
Finally got it, I think. Hope it ended up the way you wanted, it has been a bit back and forth (just learning git).

File: service.subtitles.opensubtitles/service.py [hunk 3]
```diff
@@ -128,6 +128,7 @@ def get_params(string=""):
   item['title']              = normalizeString(xbmc.getInfoLabel("VideoPlayer.OriginalTitle"))# try to get original title
   item['file_original_path'] = urllib.unquote(xbmc.Player().getPlayingFile().decode('utf-8'))# Full path of a playing file
   item['3let_language']      = [] #['scc','eng']
+  PreferredSub		     = params.get('preferredlanguage')
```
Comment (amet) [inline service.subtitles.opensubtitles/service.py]: I am doing this from my phone, how does .sort() know what PreferredSub is? its not passed up in any way

File: service.subtitles.opensubtitles/resources/lib/OSUtilities.py (hunk 2 above)
Comment (amet) [inline service.subtitles.opensubtitles/resources/lib/OSUtilities.py:75]: I have to trust you that this works as I have no easy way to test at the moment. have you tested this properly?

File: service.subtitles.opensubtitles/service.py (hunk 3 above)
Comment (goodjobswe) [inline service.subtitles.opensubtitles/service.py]: The value is assign to a variable, and we don't need to declare as global in the search function because we are only reading it. I'll do some last tests, let you know when I have.

Comment (goodjobswe) [inline service.subtitles.opensubtitles/service.py]: I did some more tests on the latest stable and on the alpha, also on the latest nightly to test the preferred language sorting and everything is working fine. I have tested everything I could think of and no errors, and everything is displaying the way it's intended to.

But I will do just a couple more, hold on a while.

Edit: Done some more tests, can't find anything off, but as I said in the other comment I am not that experienced so it would be great if someone else could have a look at it.

File: service.subtitles.opensubtitles/resources/lib/OSUtilities.py (hunk 2 above)
Comment (goodjobswe) [inline service.subtitles.opensubtitles/resources/lib/OSUtilities.py:75]: I am gonna do some more tests here

Comment (goodjobswe) [inline service.subtitles.opensubtitles/resources/lib/OSUtilities.py:75]: Now I am pretty confident everything works as intended.

This part really removes all the wrong subtitles that shows up using the old method. For example when using this method combining moviehash with IMDBid you only get subtitles for the correct movie, and if that fails you get the old result as a backup. Might need some more testing to find out if search['data'] is empty if the search is successful but don't yield any results?

Using the old method searching for subtitles for example The Other Woman you get subtitles for TVshows like: Monk, Lost, Samantha Who, The Closer, What I Like About You, Mad Men, Scandal and so on. With this method, none of that.

Might need another day on that cause I don't have more time today.

Comment (goodjobswe) [inline service.subtitles.opensubtitles/resources/lib/OSUtilities.py:75]: I managed to do some more tests, appears to be empty according to documents and further testing so should be good to go. But I am not that experienced so if someone else could have a look that would be nice.

**Comment by amet** [comment]
seems okay, one more cosmetics if you dont mind to make it easier to read the line.

or anything that you think looks good. 

```
    search_data.sort(key=lambda x: 
                         [not x['MatchedBy'] == 'moviehash',
                          not os.path.splitext(x['SubFileName'])[0] == 
                            os.path.splitext(os.path.basename(urllib.unquote(xbmc.Player().getPlayingFile().decode('utf-8'))))[0],
                          not normalizeString(xbmc.getInfoLabel("VideoPlayer.OriginalTitle")).lower() in 
                            x['SubFileName'].replace('.',' ').lower(),not x['LanguageName'] == PreferredSub]
                    )
```

**Comment by goodjobswe** [comment]
Done, also thanks for the guidance. Never worked with git before so it's needed :). Did some more cleaning replacing the values into the list instead of emptying it first and then appending the data.

Did some more tests today making sure everything works, I couldn't find anything so I say it's working very well.

**Comment by amet** [comment]
thanks!
